addons/controllers/clusterbootstrap_controller.go

```diff
@@ -300,79 +343,371 @@ func (r *ClusterBootstrapReconciler) createOrPatchclusterBootstrapFromTemplate(
+
+	secretName, err := r.GetDataValueSecretNameFromBootstrapPackage(cbpkg, cluster.Namespace)
+	if err != nil {
+		// logging has been handled in GetDataValueSecretNameFromBootstrapPackage()
+		return nil, err
+	}
+	if secretName == "" {
+		r.Log.Info(fmt.Sprintf("no data values secret is needed for ClusterBootstrapPackage: %s, nothing to be created or patched on cluster %s/%s",
+			cbpkg.RefName, cluster.Namespace, cluster.Name))
+		return nil, nil
+	}
+
+	localSecret := &corev1.Secret{}
+	key := client.ObjectKey{Namespace: cluster.Namespace, Name: secretName}
+	if err = r.Get(r.context, key, localSecret); err != nil {
+		r.Log.Error(err, "unable to fetch secret", "objectKey", key)
+		return nil, err
+	}
+
+	// TODO: This logic should be moved to cloneSecretsAndProviders()
+	// https://github.com/vmware-tanzu/tanzu-framework/issues/1729
 	// Add cluster and package labels to secrets if not already present
 	// This helps us to track the secrets in the watch and trigger Reconcile requests when these secrets are updated
-	patchedSecret := secret.DeepCopy()
-	if patchSecretWithLabels(patchedSecret, pkg.RefName, cluster.Name) {
-		if err := r.Patch(r.context, patchedSecret, client.MergeFrom(secret)); err != nil {
-			log.Error(err, "unable to patch secret labels for ", "secret", secret.Name)
+	patchedSecret := localSecret.DeepCopy()
+	if patchSecretWithLabels(patchedSecret, cbpkg.RefName, cluster.Name) {
+		if err := r.Patch(r.context, patchedSecret, client.MergeFrom(localSecret)); err != nil {
+			r.Log.Error(err, "unable to patch secret labels for ", "secret", localSecret.Name)
 			return nil, err
 		}
-		log.Info("Patched secrets with package and cluster labels to watch for changes")
+		r.Log.Info(fmt.Sprintf("patched the secret %s/%s with package and cluster labels", localSecret.Namespace, localSecret.Name))
 	}
 
-	// Now prepare the dataValuesSecret to send to target cluster
-	dataValuesSecret := &corev1.Secret{}
-	dataValuesSecret.Name = fmt.Sprintf("%s-%s-data-values", cluster.Name, packageShortName(pkg.RefName))
-	dataValuesSecret.Namespace = constants.TKGSystemNS
-	dataValuesSecret.Type = corev1.SecretTypeOpaque
+	packageRefName, _, err := util.GetPackageMetadata(r.context, r.Client, cbpkg.RefName, cluster.Namespace)
+	if err != nil {
+		r.Log.Error(err, fmt.Sprintf("unable to get Package CR %s/%s for its metadata", cluster.Namespace, cbpkg.RefName))
+		return nil, err
+	}
+
+	remoteSecret := &corev1.Secret{}
+	remoteSecret.Name = util.GenerateDataValueSecretName(cluster.Name, packageRefName)
+	// The secret will be created or patched under tkg-system namespace on remote cluster
+	remoteSecret.Namespace = r.Config.SystemNamespace
+	remoteSecret.Type = corev1.SecretTypeOpaque
 
 	dataValuesSecretMutateFn := func() error {
-		dataValuesSecret.Data = map[string][]byte{}
-		for k, v := range secret.Data {
-			dataValuesSecret.Data[k] = v
+		remoteSecret.Data = map[string][]byte{}
+		for k, v := range localSecret.Data {
```

Comment:
Shouldn't you use ```patchedSecret``` cause ```localSecret``` might have been patched?

Reply:
yes, you are right.

Reply:
fixed